Vend Flutter module App.framework as a local CocoaPod pod to be installed by a host app

jmagman opened the pull request:
## Description
Previously, the add-to-app flow installed Flutter and all plugins as pods to the host app.  To access the Flutter sub-app App.framework (dylib containing the compiled dart code), it added this line to the host app's xcconfig:
```
#include "MyApp/flutterapp/.ios/Flutter/Generated.xcconfig"
```
This gave the host app access to build settings like FLUTTER_ROOT, FLUTTER_APPLICATION_PATH, LOCAL_ENGINE, etc.  This was required to set the right environment variables to be used in the xcode_backend script.
However, it overwrote the host app's SYMROOT to `${SOURCE_ROOT}/../build/ios` which caused all kinds of horrible issues since that directory didn't exist relative to the host app's SOURCE_ROOT.
Additionally, the host app developer needed to add a script build phase to run `xcode_backend.sh build` and `xcode_backend.sh embed`.

This change:
1. Vends the App.framework as a local pod, just like Flutter.framework.  CocoaPods then handles thinning and embedding that framework into the host app, so no more need for `xcode_backend.sh embed`
1. Still calls `xcode_backend.sh build` from a new generated script with all the necessary environment variables.  Add the build script automatically to the user target (host app).  No Flutter build settings are needed by or available to the host app.
1. No more Podfile post_install hook (see https://github.com/flutter/flutter/pull/32791).  (Disabling bitcode in just the host app seemed to just work...?)
1. podhelper.rb no longer assumes directories in in .ios (could be in editable ios directory).

## Related Issues
Fixes https://github.com/flutter/flutter/issues/35901
Fixes https://github.com/flutter/flutter/issues/33094

Relevant to:
https://github.com/flutter/flutter/issues/33121
https://github.com/flutter/flutter/issues/28849
https://github.com/flutter/flutter/issues/24366

Will supersede:
https://github.com/flutter/flutter/pull/30966
https://github.com/flutter/flutter/pull/32791

## Tests

Updated:
packages/flutter_tools/test/general.shard/project_test.dart
packages/flutter_tools/test/general.shard/commands/create_test.dart
packages/flutter_tools/test/general.shard/ios/xcodeproj_test.dart

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [ ] No, this is *not* a breaking change.


Podfiles will need to be updated:
```
load File.join('my_flutter', '.ios', 'Flutter', 'podhelper.rb')
target 'MyApp' do
  install_all_flutter_pods 'my_flutter'
end
```
This will require an update to the add-to-app docs.
https://github.com/flutter/flutter/wiki/Add-Flutter-to-existing-apps#add-your-flutter-app-to-your-podfile

1. Podfile instructions will change.
1. Remove "Add a build phase" section since it's added automatically by calling install_all_flutter_pods https://github.com/flutter/flutter/wiki/Add-Flutter-to-existing-apps#add-a-build-phase-for-building-the-dart-code

jmagman committed:
Package App.framework as a local CocoaPod pod

jmagman commented on dev/integration_tests/ios_add2app/Podfile:
```diff
@@ -1,26 +1,16 @@
 target 'ios_add2app' do
-  eval(File.read(File.join(flutter_application_path, '.ios', 'Flutter', 'podhelper.rb')), binding)
+  install_flutter_plugin_pods(flutter_application_path)
+  install_flutter_application_pod(flutter_application_path, 'ios_add2app_flutter')
```
This could have called `install_all_flutter_pods flutter_application_path` if the application path matched the project name.  I expect most of them to match, but I added this workaround in case they don't.

jmagman replied:
Right now `install_flutter_plugin_pods` is used in the sub-project Podfile for the stand-alone Runner app to just pull in Flutter and the plugins, but not the App.framework.
BUT this question just made me reconsider this.  Why is the Runner sub-app special?  Why not let CocoaPods hande embedding Flutter and App there too?  Why the generated xcconfig?  Why add Flutter and App to the app itself?

The sub-app works if I let the App be embedded as a pod, so why not do that?  When I do I can remove:
1. All the script phases that call xcode_build.  CocoaPods handles thinning.
2. The embed build phases.
3. Flutter.framework and App.framework as files in the Xcode project

And if that's the case, is there any reason to not do that for all Flutter Runner apps, not just module versions?
@dnfield

jmagman replied:
Hm, although right now we aren't running pod install if there are no plugins, and I wouldn't want to add more CocoaPods paths...

jmagman replied:
Though solving this (not embedding the frameworks explicitly if we're in CocoaPod world, and letting CocoaPod embed and thin)  would fix the underlying issue of https://github.com/flutter/flutter/issues/20685, which is that two build phases want to output a Flutter.framework.

xster replied:
> BUT this question just made me reconsider this. Why is the Runner sub-app special? Why not let CocoaPods hande embedding Flutter and App there too? Why the generated xcconfig? Why add Flutter and App to the app itself?

I think it might be for the minor shortcut where the Runner can run for people without plugins and without installing cocoapods

jmagman commented on dev/integration_tests/ios_add2app/ios_add2app.xcodeproj/project.pbxproj:
```diff
@@ -326,19 +313,33 @@
+			shellScript = "\"${PODS_ROOT}/Target Support Files/Pods-ios_add2app/Pods-ios_add2app-frameworks.sh\"\n";
 			showEnvVarsInLog = 0;
 		};
+		CE6831C2A33C6C5C84521FDE /* [CP-User] Run Flutter Build Script */ = {
```
CocoaPods now adds this to the host app on a `pod install`

jmagman commented on packages/flutter_tools/lib/src/ios/xcodeproj.dart:
```diff
@@ -48,25 +49,92 @@ Future<void> updateGeneratedXcodeProperties({
+    useMacOSConfig: useMacOSConfig,
+  );
+
+  _updateGeneratedModuleBuildPhaseScript(
```
New script looks like
```
#!/bin/sh
# This is a generated file; do not edit or check into version control.
export FLUTTER_ROOT=/path/to/flutter
export FLUTTER_APPLICATION_PATH=/path/to/host/my_flutter
export FLUTTER_TARGET=lib/main.dart
export FLUTTER_BUILD_DIR=build
export SYMROOT=${SOURCE_ROOT}/../build/ios
export FLUTTER_BUILD_NAME=1.0.0
export FLUTTER_BUILD_NUMBER=1
"$FLUTTER_ROOT/packages/flutter_tools/bin/xcode_backend.sh" build
```

jmagman replied:
No, it gets created when the module gets created or re-created when the module is made editable.  It's available immediately.
Do you know what processes to people tack on that use these variables?  I'm curious why they would need FLUTTER_ variables.

xster replied:
> Do you know what processes to people tack on that use these variables? I'm curious why they would need FLUTTER_ variables. 

I'm not sure I understood the question :S

jmagman replied:
> If so, I think it's right but it's a breaking change (in case people tacked on other processes that depends on these variables). 

Do you have an example?  Or is that speculative?

xster replied:
It's pure speculation. I don't mean that we need to accommodate them or do anything special. Just announce it so people whose toolchain breaks don't have to dig around solving a mystery for an hour.

xster replied:
I think the remaining comment here is can this be in a template file

jmagman replied:
I believe the templates are never edited after their first creation/making the module editable unless the file is missing.  The only variables that are filled out in the template are things the tool knows at the time of creation, like the projectName.
The generated files like this new one and the old Generated.xcconfig are re-generated on every build.  FLUTTER_FRAMEWORK_DIR gets updated to the right engine build based on the build configuration (debug, profile, release), and the build name and number are re-read from the manifest, etc.

jmagman replied:
And this comment made me realize that we can probably parse out the FLUTTER_APPLICATION_PATH in the podhelper just like it parsed FLUTTER_ROOT to find the xcode_backend script, so that path doesn't need to get passed in from the host Podfile.

jmagman replied:
Oh never mind, it's needed to find the podhelper script in the first place.

xster replied:
Ah interesting. I never thought about the distinction. Could be worth adding a comment here.

jmagman commented on packages/flutter_tools/templates/module/ios/library/Flutter.tmpl/podhelper.rb:
```diff
@@ -1,75 +1,112 @@
+    # Fake an App.framework to have something to link against if the xcode backend script has not run yet.
+    # CocoaPods will not generate a install framework command if the vendored framework binary does not exist.
+    FileUtils.mkdir_p(app_framework_dir)
+    `echo "static const int Moo = 88;" | xcrun clang -x c -dynamiclib -o "#{app_framework_dylib}" -`
```
Matches the hack in xcode_backend.

xster replied:
😂TIL... that's wild

jmagman replied:
I actually came up with an almost-identical hack independently until I saw the one in xcode_backend.  CocoaPods just will not add the command to install a framework unless there's a dylib present at pod install, even before the project can be built.

jmagman committed:
Remove VERBOSE_SCRIPT_LOGGING

jmagman committed:
Remove another ephermeral .ios directory reference

jmagman committed:
Handle spaces

dnfield reviewed (approved):
This overall LGTM, but we should look at seeing if we could contact Add2App customers and getting a sense for whether this would be a hard breaking change for them to work with, or whether it just doesn't matter.

@xster probably has better thoughts on that

jmagman committed:
Revert "Remove another ephermeral .ios directory reference"

This reverts commit 6cffead4670e4ef47d24e4276597861786790623.

jmagman committed:
Migrate ios_host_app Podfile and project

jmagman committed:
De-CocoaPod ios_host_app to make it closer to a real add-to-app scenario, remove Generated.xcconfig include

jmagman commented:
Migration example: ios_host_app integration test.

- Removed the build phase that embedded the App.framework to let CocoaPods do it: https://github.com/flutter/flutter/pull/36793/files#diff-b5040d0f63421db53d4c954a153d7a02L20
- Removed the `xcode_backend.sh` script build phase: https://github.com/flutter/flutter/pull/36793/files#diff-b5040d0f63421db53d4c954a153d7a02L218

xster reviewed (commented):
Sorry I kept getting distracted so I wanted to send out half of my review first

xster commented on dev/integration_tests/ios_host_app/Config/Debug.xcconfig:
```diff
@@ -1,2 +0,0 @@
-#include "Flutter.xcconfig"
```
YEEEEAAAAAAAAAHHHH!!!!!

xster commented on packages/flutter_tools/templates/module/ios/library/Flutter.tmpl/podhelper.rb:
```diff
@@ -1,75 +1,112 @@
+  if !File.exist?(engine_dir)
     # Copy the debug engine to have something to link against if the xcode backend script has not run yet.
-    debug_framework_dir = File.join(flutter_root(flutter_application_path), 'bin', 'cache', 'artifacts', 'engine', 'ios')
+    # CocoaPods will not generate a install framework command if the vendored framework does not exist.
```
not sure I understood this comment

xster commented on packages/flutter_tools/templates/module/ios/library/Flutter.tmpl/podhelper.rb:
```diff
@@ -1,75 +1,112 @@
+  app_framework_dylib = File.join(app_framework_dir, 'App')
+  if !File.exist?(app_framework_dylib)
+    # Fake an App.framework to have something to link against if the xcode backend script has not run yet.
+    # CocoaPods will not generate a install framework command if the vendored framework binary does not exist.
```
Oh I think I get it now. This is just putting a fake stub into what pod install would put into the outer xcode project which gets fulfilled later?

jmagman replied:
Yup!

xster commented on packages/flutter_tools/templates/module/ios/library/Flutter.tmpl/podhelper.rb:
```diff
@@ -1,75 +1,112 @@
+  project_directory_pathname = Pathname.new Dir.pwd
+  relative = current_directory_pathname.relative_path_from project_directory_pathname
+
+  flutter_build_script_path = File.join('${SRCROOT}', relative, 'flutter_build_phase.sh');
```
I assume we'd have to document in the wiki that if your outer project and flutter module are source controlled separately, your other team members who check them out have to let them be in the same relative paths (or create symlinks)

jmagman replied:
The relative links here are being generated as the difference from where the outer host Xcode project is (Dir.pwd) and the current directory of this podhelper.rb file (`__dir__`).
Though your comment is true anyway. The Podfile needs to be checked in, which needs to pass in the path where the flutter module is located, even with the current setup: https://github.com/flutter/flutter/wiki/Add-Flutter-to-existing-apps#add-your-flutter-app-to-your-podfile

jmagman replied:
Oh I see what you mean.  The relative strings will be checked into source control since they become the inputs into the build phase.  I could change this to :input_file_list_paths and pass in a generated xcfilelist, but that feature is only in Xcode 10.

xster replied:
Right, the relative-ness of the outer project and the flutter module might not be the same for all the people in a team who check these 2 gits out. We probably can't make everything 100% magical but it would be nice to account for that scenario and minimize the git dirty-ness (possibly by best practice and documentation)

xster commented on packages/flutter_tools/templates/module/ios/library/Flutter.tmpl/projectName.podspec.tmpl:
```diff
@@ -0,0 +1,13 @@
+  s.source                = { :path => '.' }
+  s.ios.deployment_target = '8.0'
+  s.vendored_frameworks   = 'App.framework'
+  s.dependency 'Flutter'
```
Here we're assuming that the engine is the one coming from the cocoapods master spec repo? Or should this point to a specific local path?

jmagman replied:
podspecs don't let you specify a path.  It's up to the Podfile to do that.  Which is part of the reason we should take control of the Flutter CocoaPod.

xster replied:
How does this work exactly? So the outer project's Podfile went to a specific directory and found the the App.framework's pod because your install_flutter_application_pod function told it where to look. Then how does the App.framework find specifically where the matching version of the Flutter.framework comes from? 

It seems like it's coming from this line but wouldn't this just go to the online pod spec github repo to find the published engine?

jmagman replied:
As far as I can tell, the Podfile is the driver, and the dependencies between the podspecs are resolved after all the explicit pods are installed.  The podspec doesn't try to do anything, it just describes how the targets should be made dependent on each other in the build phases.  So once all the pods are installed for a target, if there are missing pods, they will be queried and fetched from the CocoaPods' repo.
So because there's a `pod 'Flutter', :path => ` in the Podfile, nothing additional is downloaded because the App podspec dependency is satisfied.
![Screen Shot 2019-07-25 at 1 53 40 PM](https://user-images.githubusercontent.com/682784/61908058-a9bc2600-aee3-11e9-816a-499e174d7d01.png)

xster replied:
Ah doh, I missed that line. Thanks for explaining. 

Can we add a comment here saying this acts as a requirement rather than a fulfillment mechanism and that it shouldn't do anything if podhelper does what it should do.

jmagman replied:
I think that's just what a podspec is, CoocaPods has pretty good docs for all this.

jmagman commented on dev/integration_tests/ios_add2app/Podfile:
```diff
@@ -1,26 +1,16 @@
 target 'ios_add2appTests' do
-  pod 'Flutter', :path => engine_dir
   inherit! :search_paths
+  install_flutter_plugin_pods(flutter_application_path)
```
This test target needs to embed the engine and the plugins.  It's not a transative dependency, this method is emplicitly installing the pod
```
pod 'Flutter', :path => engine_dir, :inhibit_warnings => true
```
I can rename.  But based on my comment about embedding the App.framework as a pod in the Runner project I think I can combine the two methods into one.

xster replied:
<I'm just curious at this point / no actions needed>what does this test do that needs the plugins?</>

jmagman replied:
You're right, I think it only needs the engine.  I will make that adjustment.

jmagman commented on dev/integration_tests/ios_add2app/ios_add2app.xcodeproj/project.pbxproj:
```diff
@@ -209,7 +213,7 @@
 			buildPhases = (
 				4375EE880A727341E9C9A57D /* [CP] Check Pods Manifest.lock */,
-				24E221E321A28B53008ADF09 /* ShellScript */,
+				CE6831C2A33C6C5C84521FDE /* [CP-User] Run Flutter Build Script */,
```
These [CP*] prefixes mean CocoaPods added them automatically.  So if you wanted to integrate directly, you could copy this step to get `flutter build` to run with the right environment variables.

xster replied:
Understood

jmagman commented on dev/integration_tests/ios_add2app/ios_add2app.xcodeproj/project.pbxproj:
```diff
@@ -326,19 +313,33 @@
+			name = "[CP-User] Run Flutter Build Script";
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "\"${SRCROOT}/flutterapp/.ios/Flutter/flutter_build_phase.sh\"";
```
It gets generated in the exact same way as the Generated.xcconfig file does (as you figured out below).

jmagman commented on packages/flutter_tools/lib/src/ios/xcodeproj.dart:
```diff
@@ -48,25 +49,91 @@ Future<void> updateGeneratedXcodeProperties({
+  generatedXcodePropertiesFile.createSync(recursive: true);
+  generatedXcodePropertiesFile.writeAsStringSync(localsBuffer.toString());
+}
+
```
Good idea.

jmagman commented on packages/flutter_tools/lib/src/ios/xcodeproj.dart:
```diff
@@ -48,25 +49,91 @@ Future<void> updateGeneratedXcodeProperties({
+
+  localsBuffer.writeln('#!/bin/sh');
+  localsBuffer.writeln('# This is a generated file; do not edit or check into version control.');
+  for (String line in xcodeBuildSettings) {
```
Curious, when would a user need to modify these?  It's all FLUTTER_-type generated variables.

xster replied:
One thing I could imagine is I was add-to-app xcode running etc, then I want to try using a local engine (which seems to work today). I think you can cause that to happen by running `flutter build ios --local_engine something` in the flutter module and then running the outer project again but it's not too clear how to trigger it from the outer project.

jmagman replied:
I think should already be passed into flutter build because FLUTTER_ENGINE and LOCAL_ENGINE are being exported in this build script https://github.com/flutter/flutter/pull/36793/files#diff-36f7d4b99eeed90b423768f2148e3ab8R160, and then parsed in xcode_backend https://github.com/flutter/flutter/blob/master/packages/flutter_tools/bin/xcode_backend.sh#L131.

xster replied:
I meant the stuff will indeed be passed into the re-entrant flutter call when it's in the script, but do the users (from the perspective of looking at the outer app project) find out how to change the script (since it's generated and says do not edit).

I just meant we can add a line of doc in this script that says if you want to modify the exported vars in this script, go into the inner flutter project and build it the way you want it to first, then your subsequent outer xcode project runs will follow those vars.

jmagman replied:
They shouldn't alter the generated file, we need to be able to blow it away at any time and not care about migrations.  I'm changing the generated script to just export the environment variables.  They can change the environment variables in the build phase between that call and the xcode_backend call if they need a local engine for testing.

xster replied:
I added a line in https://github.com/flutter/flutter/issues/32862 to make sure we document how to do this

jmagman commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -260,6 +260,9 @@ abstract class XcodeBasedProject {
   File xcodeConfigFor(String mode);
 
+  /// The script to run as part of the module pod build phase.
+  File get moduleBuildPhaseScript;
```
I think we can use this in both projects.  In my opinion we shouldn't be polluting either Xcode projects build configs with environment variables that the scripts need, like FLUTTER_* or LOCAL_ENGINE.  If we use this script we can remove the Generated.xcconfig-type files.

xster replied:
ah ok, so you're saying you envision everyone (project 1 and 2) to use this script. In the full-app templates, you would just make an xcode project that doesn't leak variables anywhere and that just uses this to make the re-entrant flutter tool calls? That would be awesome.

jmagman replied:
I think I'm going to change it to just be the export environment variables so any scripts can be run after, whether it's xcode_backend or some future `flutter build` thing.

xster replied:
I think they way you had it is great. Did it break things?

jmagman replied:
Instead of this build phase being:
```
"${SRCROOT}/flutterapp/.ios/Flutter/flutter_build_phase.sh"
```
If it's:
```
"${SRCROOT}/flutterapp/.ios/Flutter/flutter_export_environment_variables.sh"
"$FLUTTER_ROOT/packages/flutter_tools/bin/xcode_backend.sh" build
```
Then we can re-use the environment variables script in other places, like in the Runner app:
```
"${SRCROOT}/flutterapp/.ios/Flutter/flutter_export_environment_variables.sh"
"$FLUTTER_ROOT/packages/flutter_tools/bin/xcode_backend.sh" build
"$FLUTTER_ROOT/packages/flutter_tools/bin/xcode_backend.sh" embed
```

xster replied:
Reusing this generated function for full-app cases makes sense. Though I think the full-apps don't run xcode_backend.sh embed. Since we own the "outer xcode project" in that case, we just always hard linked the framework in the .xcodeproj.

jmagman commented on packages/flutter_tools/templates/module/ios/library/Flutter.tmpl/podhelper.rb:
```diff
@@ -1,75 +1,112 @@
+#   install_flutter_plugin_pods 'my_flutter'
+# end
+def install_flutter_plugin_pods(flutter_application_path = nil)
+  # If this wasn't specified, assume it's two levels up from the directory of this script.
```
Wish the person who wrote this comment had done that, that would have saved me a lot of time!

xster reviewed (commented):
The general approach LGTM. 

Can we create a build test for this in module_test_ios or something?

jmagman commented on packages/flutter_tools/templates/module/ios/library/Flutter.tmpl/podhelper.rb:
```diff
@@ -1,75 +1,112 @@
+# target 'MyApp' do
+#   install_all_flutter_pods 'my_flutter'
+# end
+def install_all_flutter_pods(flutter_application_path)
```
Good idea.

jmagman commented:
> Can we create a build test for this in module_test_ios or something?

I think it's already all being exercised in that test, it caught a few bugs.

xster commented:
> This will require an update to the add-to-app docs.
https://github.com/flutter/flutter/wiki/Add-Flutter-to-existing-apps#add-your-flutter-app-to-your-podfile

We should update the wiki but also update the WIP website branch https://github.com/flutter/website/tree/staging_add-to-app when this is merged (let's chat when you're ready to start)

jmagman committed:
- Convert build phase script to just export env variables so it can be reused in other build phases
- Add comments

jmagman committed:
Split out engine pod installation in podhelper

xster commented:
LGTM

BernardGatt commented:
This definitely makes life easier to bundle flutter & app frameworks however i'm now running into issues when adding 3rd party frameworks that have native iOS plugins, more information here: https://github.com/flutter/flutter/issues/37756

xster commented:
I just thought of something. We should probably be more strongly opinionated about whether flutter_export_environment.sh should be source controlled or gitignored.

jmagman commented:
> I just thought of something. We should probably be more strongly opinionated about whether flutter_export_environment.sh should be source controlled or gitignored (I assume more likely the latter).

This was merged a few days later:
https://github.com/flutter/flutter/commit/107f36548aec1bd278823d1ac97b3fc7e148810a

xster commented:
ah ok. I'm probably only seeing it because I took a previously created project and 'upgraded' it. That's probably ok.
